### PR TITLE
Same implementation for Sparse Multiplication for aligned and unaligned arrays

### PR DESCRIPTION
**Pull request by Anipik**
Working towards https://github.com/dotnet/machinelearning/issues/1018

### CODE REVIEW

**Comment by Anipik** [comment]
@tannergooding @eerhardt can you please take a look at this ?

File: test/Microsoft.ML.CpuMath.PerformanceTests/SsePerformanceTests.cs [hunk 1]
```diff
@@ -106,5 +106,9 @@ public void MatMulX()
         [Benchmark]
         public void MatMulTranX()
             => SseIntrinsics.MatMulTran(src, src1, dst, 1000, 1000);
+
+        [Benchmark]
+        public void MatMulPX()
```
Comment (eerhardt) [inline test/Microsoft.ML.CpuMath.PerformanceTests/SsePerformanceTests.cs]: Why is this test called `MatMulPX`, when the method it is benchmarking is named `MatMulP`?

File: src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs [hunk 2]
```diff
@@ -114,7 +114,7 @@ public static void MatTimesSrc(bool tran, AlignedArray mat, AlignedArray src, Al
             else if (Sse.IsSupported)
             {
                 Contracts.Assert(crun <= dst.Size);
-                SseIntrinsics.MatMulPA(mat, rgposSrc, srcValues, posMin, iposMin, iposLim, dst, crun, srcValues.Size);
+                SseIntrinsics.MatMulP(mat, rgposSrc, srcValues, posMin, iposMin, iposLim, dst, crun, srcValues.Size);
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs]: The corresponding `AvxIntrinsics` and `SseIntrinsics` methods should be named the same. This one is `MatMulP`, while the Avx one is `MatMulPX`.

File: src/Microsoft.ML.CpuMath/Thunk.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ internal static unsafe class Thunk
         public static extern void MatMulX(/*const*/ float* pmat, /*const*/ float* psrc, float* pdst, int crow, int ccol);
 
         [DllImport(NativePath), SuppressUnmanagedCodeSecurity]
-        public static extern void MatMulPA(/*const*/ float* pmat, /*const*/ int* pposSrc, /*const*/ float* psrc,
+        public static extern void MatMulP(/*const*/ float* pmat, /*const*/ int* pposSrc, /*const*/ float* psrc,
             int posMin, int iposMin, int iposLim, float* pdst, int crow, int ccol);
         [DllImport(NativePath), SuppressUnmanagedCodeSecurity]
         public static extern void MatMulPX(/*const*/ float* pmat, /*const*/ int* pposSrc, /*const*/ float* psrc,
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/Thunk.cs]: Who calls `MatMulPX` anymore?

Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/Thunk.cs]: I also don't see it in `Sse.cpp`.  This should be removed.

---
In reply to: [225949263](https://github.com/dotnet/machinelearning/pull/1274#discussion_r225949263) [](ancestors = 225949263)

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 4]
```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
...
-            Contracts.Assert(HasCompatibleAlignment(mat));
-            Contracts.Assert(HasCompatibleAlignment(src));
-            Contracts.Assert(HasCompatibleAlignment(dst));
+            Contracts.Assert(crow % 4 == 0);
+            Contracts.Assert(ccol % 4 == 0);
 
+            MatMulP(mat.Items, rgposSrc, src.Items, posMin, iposMin, iposEnd, dst.Items, crow, ccol);
+        }
+
+        public static unsafe void MatMulP(float[] mat, int[] rgposSrc, float[] src,
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: I missed this in the last review - but can you make these `ReadOnlySpan<T>` instead of `T[]`? That way they are consistent with the rest of the methods in this file.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 5]
```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
...
-            fixed (float* pDstStart = &dst.Items[0])
-            fixed (float* pMatStart = &mat.Items[0])
+            fixed (float* psrc = &src[0])
+            fixed (float* pdst = &dst[0])
+            fixed (float* pmat = &mat[0])
             fixed (int* pposSrc = &rgposSrc[0])
+            fixed (uint* pLeadingAlignmentMask = &LeadingAlignmentMask[0])
+            fixed (uint* pTrailingAlignmentMask = &TrailingAlignmentMask[0])
             {
-                float* psrc = GetAlignedBase(src, pSrcStart);
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: I don't believe there are any more callers to `GetAlignedBase`. This method can be removed. Same for `HasCompatibleAlignment`.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 6]
```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
...
+                            float* pm3 = pm2 + ccol;
+                            Vector128<float> result = Sse.SetZeroVector128();
+
+                            int* ppos = pposMin;
 
-                        ppos++;
+                            while (ppos < pposEnd)
+                            {
+                                int col = *ppos;
+                                Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: What exactly about this algorithm needs alignment?  I don't see us calling any "Aligned" methods. Is `Sse.Store(pDstCurrent, result);` supposed to be `Sse.StoreAligned(pDstCurrent, result);`?

And if that is the only thing that needs alignment, do we know that this extra complexity outweighs the perf difference?

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 7]
```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
...
+                        Vector128<float> leadingMask = Sse.LoadVector128(((float*)(pLeadingAlignmentMask)) + ((4 - remainder) * 4));
+
+                        float* pm1 = pm0 + ccol;
+                        float* pm2 = pm1 + ccol;
+                        float* pm3 = pm2 + ccol;
+                        Vector128<float> result = Sse.SetZeroVector128();
+
+                        int* ppos = pposMin;
+
+                        while (ppos < pposEnd)
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Can we refactor this inner loop into its own method? It is duplicated 3 times in this method.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs [hunk 8]
```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
...
         // Partial sparse source vector.
-        public static unsafe void MatMulPA(AlignedArray mat, int[] rgposSrc, AlignedArray src,
-                                        int posMin, int iposMin, int iposEnd, AlignedArray dst, int crow, int ccol)
+        public static unsafe void MatMulP(AlignedArray mat, int[] rgposSrc, AlignedArray src,
+                                int posMin, int iposMin, int iposEnd, AlignedArray dst, int crow, int ccol)
         {
-            Contracts.Assert(HasCompatibleAlignment(mat));
-            Contracts.Assert(HasCompatibleAlignment(src));
-            Contracts.Assert(HasCompatibleAlignment(dst));
+            Contracts.Assert(crow % 4 == 0);
```
Comment (eerhardt) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: I think these asserts should move into the method below (the one taking in `float[] mat, int[] rgposSrc`). That method is the one that requires `ccol` to be a multiple of 4.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 6 above)
Comment (tannergooding) [inline src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: note to self: I want to check the codegen of this and ensure that it is being emitted "optimally" (two loads and three unpack with two folded loads; rather than as four loads and three unpack).

Comment (tannergooding) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Actually, looking at the algorithm, this isn't doing anything where the alignment should matter, since it is operating on sparse data (and has to do 4 float loads, rather than a single 16-byte load).

Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: The only one that I identified was the `Sse.Store(pDstCurrent, result)`.

Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Ok, this makes sense to me. Thanks for the explanation.

Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Thinking a bit more - do we want to change the other `MatMul` algorithm then? That one is aligning the input "matrix", and not the resulting destination.
Although, we are doing `(4 * ccol)` aligned loads for every 1 store. So maybe it is best to align that many loads?

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 4 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: will it have any effect on performance ?

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 6 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: > The only one that I identified was the Sse.Store(pDstCurrent, result)

yes earlier it was doing one alignedLoad(it was removed while removing the add flag) and alignedStore.

Comment (Anipik) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: > Although, we are doing (4 * ccol) aligned loads for every 1 store. So maybe it is best to align that many loads?

yes that was motivation behind the aligning the input matrix because it has larger number of loads.

File: src/Microsoft.ML.CpuMath/Thunk.cs (hunk 3 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/Thunk.cs]: i think its called in ```Avx.cs```. but i will remove it because ```Avx.cs```  file is not used anywhere

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 4 above)
Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Measure it. But from everything I've measured in the past, my guess would be "no"

File: src/Microsoft.ML.CpuMath/Thunk.cs (hunk 3 above)
Comment (eerhardt) [reply src/Microsoft.ML.CpuMath/Thunk.cs]: Yes, please remove it from `Avx.cs` as well.

File: src/Microsoft.ML.CpuMath/SseIntrinsics.cs (hunk 4 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/SseIntrinsics.cs]: Yeah you are right, the performance remains same

File: src/Native/CpuMathNative/CMakeLists.txt [hunk 9]
```diff
@@ -2,15 +2,13 @@ project (CpuMathNative)
 
 set(SOURCES
     Sse.cpp
-    Avx.cpp
     MathLinux.S
 )
 
 if(WIN32)
-    set_property(SOURCE Avx.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " /arch:AVX")
+    set_property(SOURCE Sse.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " -mavx")
```
Comment (eerhardt) [inline src/Native/CpuMathNative/CMakeLists.txt]: Why is this necessary? It wasn't before, was it?

Also, we should follow the typical windows style compile flags: `/arch`.

File: test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs [hunk 10]
```diff
@@ -116,7 +116,7 @@ public void MatMulTranTest(int matTest, int srcTest, int dstTest, float[] expect
         [InlineData(0, 0, 0, new float[] { 38.25002f, 38.25002f, 38.25002f, 38.25002f, 38.25002f, 38.25002f, 38.25002f, 38.25002f })]
         [InlineData(1, 1, 0, new float[] { 910f, 2190f, 3470f, 4750f, 6030f, 7310f, 8590f, 9870f })]
         [InlineData(1, 0, 1, new float[] { 95f, 231f, 367f, 503f, 639f, 775f, 911f, 1047f, 1183f, 1319f, 1455f, 1591f, 1727f, 1863f, 1999f, 2135f })]
-        public void MatMulPATest(int matTest, int srcTest, int dstTest, float[] expected)
+        public void MatMulPTest(int matTest, int srcTest, int dstTest, float[] expected)
```
Comment (eerhardt) [inline test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs]: Why isn't this test called `MatTimesSrcTest` to match the function it is testing?

File: src/Native/CpuMathNative/CMakeLists.txt (hunk 9 above)
Comment (tannergooding) [reply src/Native/CpuMathNative/CMakeLists.txt]: I agree, we should be using `/arch:AVX` for MSVC and `-mavx` for GCC/Clang

**Comment by eerhardt** [comment]
> #endif

It looks like this whole block can be removed as well. It is no longer used.

---
Refers to: src/Native/CpuMathNative/Sse.cpp:79 in db669af. [](commit_id = db669afaf5d8f44aa3a7361a952dbee67f6b8afb, deletion_comment = False)

File: test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs (hunk 10 above)
Comment (Anipik) [reply test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs]: A couple of overload are present for ```MatTimesSrc``` . And these overloads also contains flags such as ```trans``` in order to decide what version of matrix multiplication will finally be executed. 
So we are using this name for test to show which matrix multiplication is being used

File: src/Native/CpuMathNative/CMakeLists.txt (hunk 9 above)
Comment (Anipik) [reply src/Native/CpuMathNative/CMakeLists.txt]:
```
if(WIN32)
    set_property(SOURCE Sse.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " -mavx")
 else()
    set_property(SOURCE Sse.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " -msse3")
    set_property(SOURCE Sse.cpp APPEND_STRING PROPERTY COMPILE_FLAGS " -mavx")
endif
```
is this okay ? just changed the file Avx.cpp to Sse.cpp for all compile flags as Avx.cpp file is no longer present.

Comment (tannergooding) [reply src/Native/CpuMathNative/CMakeLists.txt]: It is slightly more complicated than that. If we compile using `/arch:AVX` and `-mavx` then it will compile with the VEX encoding, which will only be supported on (most) processors from 2011 or later.

Comment (Anipik) [reply src/Native/CpuMathNative/CMakeLists.txt]: so what should this file look like if we remove Avx.cpp as funtions from this file is not used anywhere

File: test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs (hunk 10 above)
Comment (eerhardt) [reply test/Microsoft.ML.CpuMath.UnitTests.netcoreapp/UnitTests.cs]: I think this test should be named `MatTimesSrcSparse`. It would match the name of the method it is testing, and gives the indication for which overload it is representing (the sparse vector overload).

File: src/Native/CpuMathNative/CMakeLists.txt (hunk 9 above)
Comment (Anipik) [reply src/Native/CpuMathNative/CMakeLists.txt]: It works , i was using ```!``` instead of ```NOT``` which was leading to compile error earlier

Comment (tannergooding) [reply src/Native/CpuMathNative/CMakeLists.txt]: We likely want to be very explicit that ML.NET (on full framework) requires a CPU with SSE3 support (while .NET Core will only require SSE2).

Comment (Anipik) [reply src/Native/CpuMathNative/CMakeLists.txt]: so do I need to change something in this file or add some comment here ?

Comment (eerhardt) [reply src/Native/CpuMathNative/CMakeLists.txt]: No - this now looks fine.

---
In reply to: [226419431](https://github.com/dotnet/machinelearning/pull/1274#discussion_r226419431) [](ancestors = 226419431)

**Comment by Anipik** [comment]
@tannergooding did u get the chance to look at the codegen for this ?

**Comment by tannergooding** [comment]
No, not yet.

File: src/Microsoft.ML.CpuMath/Avx.cs [hunk 11]
```diff
@@ -1,1165 +0,0 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (tannergooding) [inline src/Microsoft.ML.CpuMath/Avx.cs]: nit: Removing dead code seems unrelated to the PR.

File: src/Microsoft.ML.CpuMath/AvxIntrinsics.cs [hunk 12]
```diff
@@ -333,34 +418,31 @@ public static unsafe void MatMulX(float[] mat, float[] src, float[] dst, int cro
                         int col1 = *ppos;
                         int col2 = col1 + 4 * ccol;
                         Vector256<float> x1 = Avx.SetVector256(pm3[col2], pm2[col2], pm1[col2], pm0[col2],
-                                                                pm3[col1], pm2[col1], pm1[col1], pm0[col1]);
+                                                                   pm3[col1], pm2[col1], pm1[col1], pm0[col1]);
                         Vector256<float> x2 = Avx.SetAllVector256(pSrcCurrent[col1]);
                         x2 = Avx.Multiply(x2, x1);
```
Comment (tannergooding) [inline src/Microsoft.ML.CpuMath/AvxIntrinsics.cs]: nit: Now that https://github.com/dotnet/machinelearning/pull/1292 has been merged, it would be nice to update the `Multiply, followed by Add` cases to call the `MultiplyAdd` helper.

File: src/Microsoft.ML.CpuMath/AvxIntrinsics.cs [hunk 13]
```diff
@@ -294,32 +275,136 @@ public static unsafe void MatMulX(float[] mat, float[] src, float[] dst, int cro
...
+                        float* pm3 = pm2 + ccol;
+                        Vector256<float> result = Avx.SetZeroVector256<float>();
+
+                        int* ppos = pposMin;
+
+                        while (ppos < pposEnd)
+                        {
+                            int col1 = *ppos;
+                            int col2 = col1 + 4 * ccol;
+                            Vector256<float> x1 = Avx.SetVector256(pm3[col2], pm2[col2], pm1[col2], pm0[col2],
```
Comment (tannergooding) [inline src/Microsoft.ML.CpuMath/AvxIntrinsics.cs]: Don't we have a helper method for this?

File: src/Microsoft.ML.CpuMath/Avx.cs (hunk 11 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/Avx.cs]: Some of the functions had the similar names like matmulPX in avx.cpp files which were not getting used anywhere so in order to avoid confusion I removed all the deadCode. Earlier I was delaying it until the aligned array removal PR.

File: src/Microsoft.ML.CpuMath/AvxIntrinsics.cs (hunk 13 above)
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/AvxIntrinsics.cs]: no its different , the one we have the indexs are continous 
```return Avx.SetVector256(src[idx[7]], src[idx[6]], src[idx[5]], src[idx[4]], src[idx[3]], src[idx[2]], src[idx[1]], src[idx[0]]);```

**Comment by Anipik** [comment]
``` ini
Before
```
|      Method |     Avx  |     Sse   |    Native |
|------------ |---------:|----------:|----------:|
| MatMulPX(S = 10%)     | 95.28 us | 114.00 us | 91.8  us |
| MatMulPX (S = 20%)| 146.2 us | 181.2 us | 131.3  us |


``` ini
After
```
|      Method |     Avx  |     Sse   |    Native |
|------------ |---------:|----------:|----------:|
| MatMulPX(S = 10%)     | 96.46 us | 106.3 us | 92.8  us |
| MatMulPX (S = 20%)| 140 us | 171.7 us | 131.1  us |


As the matrix becomes more dense , the new algorithm becomes faster

cc @danmosemsft @eerhardt @tannergooding

**Comment by Anipik** [comment]
@tannergooding i have resolved the conflicts and addressed the feedback

**Comment by Anipik** [comment]
I have restarted the queue and new build passed successfully

File: src/Microsoft.ML.CpuMath/AvxIntrinsics.cs [hunk 14]
```diff
@@ -312,32 +293,134 @@ public static unsafe void MatMulX(float[] mat, float[] src, float[] dst, int cro
...
+                nuint address = (nuint)(pDstCurrent);
+                int misalignment = (int)(address % 32);
+                int length = crow;
+                int remainder = 0;
+
+                if ((misalignment & 3) != 0)
+                {
+                    while (pDstCurrent < pDstEnd)
+                    {
+                        Avx.Store(pDstCurrent, SparseMultiplicationAcrossRow());
```
[the remark this answers is not in the document]
Comment (Anipik) [reply src/Microsoft.ML.CpuMath/AvxIntrinsics.cs]: i checked the perf with and without the ```SparseMultiplicationAcrossRow``` , and its the same.
do you want me to remove the function call entirely or move it out of this function ?

[the remark this answers is not in the document]
Comment (tannergooding) [reply src/Microsoft.ML.CpuMath/AvxIntrinsics.cs]: If the perf is the same, then it's fine 😄
